tests/test_defects.py

```diff
@@ -103,30 +108,472 @@ def setUp(self):
 
         self.defaultTask = cpPipe.defects.FindDefectsTask(config=self.defaultConfig)
 
-    def test_defectFindingAllSensor(self):
+        self.allDefectsList = measAlg.Defects()
+
+        self.brightDefectsList = measAlg.Defects()
+        for d in self.brightBBoxes:
+            self.brightDefectsList.append(d)
+            self.allDefectsList.append(d)
+
+        self.darkDefectsList = measAlg.Defects()
+        for d in self.darkBBoxes:
+            self.darkDefectsList.append(d)
+            self.allDefectsList.append(d)
+
+    def check_maskBlocks(self, inputDefects, expectedDefects):
+        """A helper function for the tests of maskBlocksIfIntermitentBadPixelsInColumn.
+        """
         config = copy.copy(self.defaultConfig)
+        config.badOnAndOffPixelColumnThreshold = 10
+        config.goodPixelColumnGapThreshold = 5
         config.nPixBorderUpDown = 0
         config.nPixBorderLeftRight = 0
+
+        task = cpPipe.defects.FindDefectsTask(config=config)
+
+        defectsWithColumns = task.maskBlocksIfIntermitentBadPixelsInColumn(inputDefects)
+
+        boxesMeasured = []
+        for defect in defectsWithColumns:
+            boxesMeasured.append(defect.getBBox())
+
+        for boxInput in expectedDefects:
+            self.assertIn(boxInput, boxesMeasured)
```

Comment:
Could you please elaborate on this? I think that the test needs to check that the expected defects are part of the set of measured defects, without requiring them to have them the same cardinality necessarily. In any case, I tried to implement the second option that you suggest "self.assertEqual(len(boxInput, boxesMeasured)", but I'm a bit confused because "boxInput" is just an element of the set  "expectedDefects".

Reply:
You should know exactly what defects should be produced by each test, without that you miss things like the test errors I described below (where the "blob" was broken up but you didn't check for the second part of it).

In both of my examples, replace `boxInput` with `expectedDefects`: I've updated the comment. Sadly, `Defect`s are not sortable, so my `sorted` test wouldn't work (and I think we want sorting there, because the order of the output shouldn't matter).

Reply:
I think that each of the tests added for this ticket should check that the set of expected defects in each test is contained entirely within the set of measured defects, which will be a set of defects found by the general defect finding code and by the extra columnar code added in this branch. As a consequence of this,  the extra condition that you suggest be added, `self.assertEqual(len(expectedDefects), len(boxesMeasured))`, will not be satisfied (I assumed that this is what you meant since in your comment I see `self.assertEqual(len(expectedDefects), boxesMeasured)`, but please correct me if I am wrong).  The `boxesMeasured` set will contain more elements than the `expectedDefects` set.  For example, in the test where we have bad pixels every other pixel along the whole column (`test_maskBlocks_every_other_pixel_bad_greater_than_threshold`), the defect-finding code that runs before the columnar code of this ticket will first find the intermittent bad pixels (and also the bad pixels defined for other tests in the code, such as the ones in lines 73 and 75 ), then, the columnar code of this branch will mark the whole column as a bad region because the two conditions for doing so are triggered (namely, (a) the number of bad pixels in the column is greater than the threshold1=10, and (b) (given (a)  is satisfied), the number of consecutive good pixels is not greater than threshold2=5). This is indeed what happens when I run the code for this test: 

`EXPECTED:  [Box2I(minimum=Point2I(50, 10), dimensions=Extent2I(1, 31))]

MEASURED:  [Box2I(minimum=Point2I(0, 15), dimensions=Extent2I(3, 3)), Box2I(minimum=Point2I(100, 123), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(5, 0), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(7, 62), dimensions=Extent2I(2, 2)), Box2I(minimum=Point2I(50, 10), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 12), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 14), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 16), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 18), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 20), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 22), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 24), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 26), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 28), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 30), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 32), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 34), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 36), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 38), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 40), dimensions=Extent2I(1, 1)), Box2I(minimum=Point2I(50, 10), dimensions=Extent2I(1, 31))]`

I think that the critical thing is to make sure that all the defects in `EXPECTED` (in this case only one: `Box2I(minimum=Point2I(50, 10), dimensions=Extent2I(1, 31))`), are found in `MEASURED~, which is the case, and that's why the condition is 

`for boxInput in expectedDefects:
            self.assertIn(boxInput, boxesMeasured)`

The errors that you found in the previous comment had to do with an incomplete `EXPECTED` set (this set is determined by hand knowing what we would want the code in this branch to do, and hence the images in Jira ticket DM-19903 to aid in this), but they did not indicate that the code was not working (the code worked, it just I forgot to check for a few more extra boxes).

Reply:
Hi both —

Since I understand this issue is blocking this ticket being reviewed, we need to resolve it.

As I understand it, there are two issues here:

- @parejkoj is correct that we should test that `maskBlocksIfIntermitentBadPixelsInColumn` doesn't mask *more* than it is supposed to. By my (superficial) reading of the code, that's certainly a possibility — `boxesMeasured` could contain boxes that cover effectively the whole image in addition to the `expectedDefects` and the test would pass. I agree with him that this is less than ideal.

- @plazas is correct that `boxesMeasured` doesn't just contain the columnar defects, but *also* the inputs to `maskBlocksIfIntermittentBadPixelsInColumn`. I therefore don't think that checking for equality between `boxesMeasured` and `expectedDefects` is helpful.

Given that, I think the right checks are:

- Iterate over `boxesMeasured`, and, for *every* entry, ensure that it appears in *either* `inputDefects` *or* `expectedDefects`.
- Iterate over `expectedDefects`, and, for *every* entry, ensure that it appears in `boxesMeasured` (this is what's already happening).

I think that's just another extra couple of lines of code.

Reply:
Thank you, @jdswinbank. I have implemented your suggestions. @parejkoj, please let me know if things look good for you.

Reply:
(and the tests pass)